bpo-41073: PyType_GetSlot() can now accept static types.

shihai1991 opened the pull request:
PyType_GetSlot() can now accept static types.

Co-Authored-By: Petr Viktorin <encukou@gmail.com>

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-41073](https://bugs.python.org/issue41073) -->
https://bugs.python.org/issue41073
<!-- /issue-number -->

Automerge-Triggered-By: GH:encukou

shihai1991 committed:
PyType_GetSlot() could accept static types

shihai1991 committed:
update docs' style

shihai1991 committed:
use typeslots not in limit apis

shihai1991 committed:
add test cases to against PyType_GetSlot()

shihai1991 committed:
update typeslots.h

encukou commented on Include/typeslots.h:
```diff
@@ -88,3 +88,43 @@
+#define Py_tp_vectorcall_offset 95
+#define Py_tp_version_tag 96
+#define Py_tp_weaklist 97
+#define Py_tp_weaklistoffset 98
```
This should not be a part of [bpo-41073](https://bugs.python.org/issue41073). Can you do this in a different issue/PR, preferably with a discussion for each of the newly added slot?
I would not want to expose the `Py_tp_as_*` slots at all; their contents are covered in other slots that should be used instead. And slots like `Py_tp_version_tag` are specific to CPython; I don't think they should be exposed in the general C API.

shihai1991 replied:
> I would not want to expose the `Py_tp_as_*` slots at all; their contents are covered in other slots that should be used instead. And slots like `Py_tp_version_tag` are specific to CPython; I don't think they should be exposed in the general C API.

Hm, make sense. I am not sure the scope of extension. I can update it later.

shihai1991 replied:
> This should not be a part of [bpo-41073](https://bugs.python.org/issue41073). Can you do this in a different issue/PR, preferably with a discussion for each of the newly added slot?

Hi, petr. I have created another bpo: https://bugs.python.org/issue41618

encukou reviewed (changes requested):
Can you also test getting `Py_nb_add` and `Py_mp_length` from `PyLong_Type`?
That will expose the reason this is complicated: static types don't have the sub-slots structs allocated as part of the type.

shihai1991 committed:
Add generate statictypeslots feature

shihai1991 committed:
revert typeslots.inc to master base

shihai1991 committed:
update get static type slots

shihai1991 committed:
Merge remote-tracking branch 'origin/master' into bpo_41073_add_typeslots1

* origin/master: (63 commits)
  bpo-41627: Distinguish 32 and 64-bit user site packages on Windows (GH-22098)
  bpo-38585: Remove references to defusedexpat (GH-22095)
  bpo-41721: Add xlc options (GH-22096)
  bpo-40486: Specify what happens if directory content change diring iteration (GH-22025)
  bpo-41638: Improve ProgrammingError message for absent parameter. (GH-21999)
  bpo-41713: _signal doesn't use multi-phase init (GH-22087)
  bpo-41700: Skip test if the locale is not supported (GH-22081)
  [doc] Update documentation on logging optimization. (GH-22075)
  Fix 'gather' rules in the python parser generator (GH-22021)
  bpo-41697: Correctly handle KeywordOrStarred when parsing arguments in the parser (GH-22077)
  [doc] Fix a typo in the graphlib docs (#22030)
  bpo-1635741: Port _signal module to multi-phase init (PEP 489) (GH-22049)
  bpo-39883: Use BSD0 license for code in docs (GH-17635)
  bpo-39010: Improve test shutdown (#22066)
  bpo-41696: Fix handling of debug mode in asyncio.run (#22069)
  bpo-41690: Use a loop to collect args in the parser instead of recursion (GH-22053)
  closes bpo-41689: Preserve text signature from tp_doc in C heap type creation. (GH-22058)
  Fix invalid escape sequences in the peg_highlight Sphinx extension (GH-22047)
  bpo-41675: Modernize siginterrupt calls (GH-22028)
  bpo-41685: Don't pin setuptools version anymore in Doc/Makefile (GH-22062)
  ...

shihai1991 committed:
Fix compile error

shihai1991 committed:
Use petr's idea

encukou commented on Include/typeslots.h:
```diff
@@ -88,3 +88,51 @@
+#define Py_tp_nb_or 108
+#define Py_tp_nb_int 109
+#define Py_tp_nb_reserved 110
+#define Py_tp_nb_float 111
```
What's the difference between `Py_tp_nb_float` and `Py_nb_float` (11)?

shihai1991 replied:
Oh, no difference. I copy this slots from old commit~.

shihai1991 replied:
Oh, wait~
`PyHeapTypeObject` have `as_number` and `PyTypeObject` have  `tp_as_number`, so it is need  two vars, no?

encukou replied:
You can use ` tp_as_number` in both cases.
`PyHeapTypeObject` also have `as_number` directly included, but ` tp_as_number` points to it.

See: https://github.com/python/cpython/blob/345cd37abe324ad4f60f80e2c3133b8849e54e9b/Objects/typeobject.c#L2984-L2989

shihai1991 replied:
oh, you are right.

encukou commented on Include/typeslots.h:
```diff
@@ -88,3 +88,51 @@
+#define Py_tp_mro 89
+#define Py_tp_cache 90
+#define Py_tp_subclasses 91
+#define Py_tp_weaklist 92
```
Please don't add slots for data (as opposed to functions).

shihai1991 replied:
make sense. I will update it tomorrow.

encukou commented on Modules/_testcapimodule.c:
```diff
@@ -1018,6 +1018,42 @@ test_buildvalue_N(PyObject *self, PyObject *Py_UNUSED(ignored))
+    assert(tp_call == 0);
+
+    binaryfunc nb_add = PyType_GetSlot(&PyLong_Type, Py_tp_nb_add); 
+    PyObject *result = nb_add(object, object);
```
You can rather assert somethiong like `nb_add == PyLong_Type->tp_as_number->nb_add`.

shihai1991 replied:
no problem.

encukou commented on Objects/typeobject.c:
```diff
@@ -2882,6 +2888,284 @@ static const short slotoffsets[] = {
 #include "typeslots.inc"
 };
 
+static const PySlot_Offset pyslot_offsets[] = {
```
This structure should be generated automatically. 
Did you write it by hand, or do you have a script for it?

shihai1991 replied:
I write it by hand temporarily.

shihai1991 replied:
petr. How about store this `pyslot_offsets` to an seperate internal seperate header?
1. It's too long~
2. more easy for script to generate.

encukou replied:
Yes!

It would be best to change the existing `typeslots.inc` rather than adding a new file:
- Remove `slot` from `PySlot_Offset` (the slots should be identified by their position in the array, like the current `slotoffsets`
- Change the existing `slotoffsets` to be an array of `PySlot_Offset`
- Change `typeslots.py` to generate `PySlot_Offset` structures instead of just numbers.
- Change all existing uses of `slotoffsets` to use `PySlot_Offset` (follow the two offsets, even for heap types)

encukou commented on Objects/typeobject.c:
```diff
@@ -3119,15 +3403,24 @@ PyType_FromSpec(PyType_Spec *spec)
-        /* Extension module requesting slot from a future version */
-        return NULL;
+
+    for (cur_offset = pyslot_offsets; cur_offset->slot; cur_offset++) {
```
It would be better to index the array directly; slot numbers are small integers.

shihai1991 committed:
apply petr's comment

shihai1991 committed:
Merge remote-tracking branch 'origin/master' into bpo_41073_add_typeslots1

* origin/master: (365 commits)
  bpo-42029: Remove IRIX code (GH-23023)
  bpo-42143: Ensure PyFunction_NewWithQualName() can't fail after creating the func object (GH-22953)
  bpo-34204: Use pickle.DEFAULT_PROTOCOL in shelve (GH-19639)
  bpo-41805: Documentation for PEP 585 (GH-22615)
  bpo-42161: Micro-optimize _collections._count_elements() (GH-23008)
  bpo-42161: Remove private _PyLong_Zero and _PyLong_One (GH-23003)
  bpo-42099: Fix reference to ob_type in unionobject.c and ceval (GH-22829)
  bpo-41659: Disallow curly brace directly after primary (GH-22996)
  bpo-6761: Enhance __call__ documentation (GH-7987)
  bpo-42161: Modules/ uses _PyLong_GetZero() and _PyLong_GetOne() (GH-22998)
  bpo-41474, Makefile: Add dependency on cpython/frameobject.h (GH-22999)
  bpo-42157: Rename unicodedata.ucnhash_CAPI (GH-22994)
  bpo-42161: Use _PyLong_GetZero() and _PyLong_GetOne() (GH-22995)
  bpo-30681: Support invalid date format or value in email Date header (GH-22090)
  bpo-42161: Add _PyLong_GetZero() and _PyLong_GetOne() (GH-22993)
  bpo-42123: Run the parser two times and only enable invalid rules on the second run (GH-22111)
  bpo-42157: Convert unicodedata.UCD to heap type (GH-22991)
  bpo-42157: unicodedata avoids references to UCD_Type (GH-22990)
  bpo-39101: Fixes BaseException hang in IsolatedAsyncioTestCase. (GH-22654)
  bpo-1635741: _PyUnicode_Name_CAPI moves to internal C API (GH-22713)
  ...

shihai1991 committed:
update release doc

shihai1991 commented on Include/typeslots.h:
```diff
@@ -88,3 +88,6 @@
 /* New in 3.5 */
 #define Py_tp_finalize 80
 #endif
+#define Py_tp_dict 81
```
Not sure do we need it or not

shihai1991 replied:
adding new slots is not this PR's work, delete it.

shihai1991 committed:
remove new adding slots

encukou commented on Include/typeslots.h:
```diff
@@ -88,51 +88,6 @@
-#define Py_tp_nb_inplace_matrix_multiply 128
+#define Py_tp_dict 81
+#define Py_tp_mro 82
+#define Py_tp_subclasses 83
```
Please don't add new slots in this PR.

encukou replied:
Oops, never mind – I must have looked at an older commit when I wrote this.

encukou commented on Objects/typeobject.c:
```diff
@@ -57,8 +57,8 @@ static unsigned int next_version_tag = 0;
 
 typedef struct PySlot_Offset {
     int slot;
```
`pyslot_offsets[i].slot` should always be `i`, so it should be possible to remove `slot`.

shihai1991 replied:
You are right, have been updated, thanks.

encukou reviewed (commented):
I didn't have time for a thorough review yet, but the changes look great!
I have ~~two~~ one more comment now:

shihai1991 committed:
apply petr's comment

encukou committed:
Call test_get_statictype_slots in the test suite

encukou committed:
Raise exception on the invalid slot 0

encukou committed:
Use if rather than assert in tests

C asserts are removed in non-debug builds.

encukou committed:
Add/adjust comments & news entries

encukou committed:
Align method definitions in _testcapi

This is how most of the table is formatted.
Fix a few lines around the one added.

encukou committed:
Rewrite the slot-setting code a bit

IMO, things are more easier to follow this way (combined with
the comment added to pyslot_offsets).

encukou committed:
Remove the loop in PyType_GetSlot; we already know the position

encukou committed:
typeslots.py: Add back code writing null entries

This way the table will be kept in sync if a slot is removed
(which shouldn't happen).

shihai1991 commented on Lib/test/test_capi.py:
```diff
@@ -544,6 +544,10 @@ def test_pynumber_tobase(self):
         self.assertRaises(SystemError, pynumber_tobase, 123, 0)
 
+    def test_get_statictype_slots(self):
+        from _testcapi import test_get_statictype_slots
```
Why do we need add this testcase in python level?  Test_testcapi will take care of it:)

encukou replied:
Ah, you're right!

shihai1991 committed:
Merge remote-tracking branch 'origin/master' into bpo_41073_add_typeslots1

* origin/master: (80 commits)
  bpo-42282: Fold constants inside named expressions (GH-23190)
  bpo-41028: Doc: Move switchers to docsbuild-scripts. (GH-20969)
  bpo-42133: update parts of the stdlib to fall back to `__spec__.loader` when `__loader__` is missing (#22929)
  Remove outdated reference to pywin32 from platform module (GH-22005)
  bpo-41832: PyType_FromModuleAndSpec() now accepts NULL tp_doc (GH-23123)
  Minor grammar edits for the descriptor howto guide (GH-#23175)
  bpo-42179: Doc/tutorial: Remove mention of __cause__ (GH-23162)
  bpo-26389: Allow passing an exception object in the traceback module (GH-22610)
  bpo-42260: PyConfig_Read() only parses argv once (GH-23168)
  bpo-42260: Add _PyConfig_FromDict() (GH-23167)
  bpo-41877 Check for asert, aseert, assrt in mocks (GH-23165)
  [docs] fix wrongly named AsyncContextDecorator (GH-23164)
  bpo-42262: Add Py_NewRef() and Py_XNewRef() (GH-23152)
  bpo-42266: Handle monkey-patching descriptors in LOAD_ATTR cache (GH-23157)
  bpo-40816 Add AsyncContextDecorator class (GH-20516)
  bpo-42260: Add _PyInterpreterState_SetConfig() (GH-23158)
  Disable peg generator tests when building with PGO (GH-23141)
  bpo-1635741: _sqlite3 uses PyModule_AddObjectRef() (GH-23148)
  bpo-1635741: Fix PyInit_pyexpat() error handling (GH-22489)
  bpo-42260: Main init modify sys.flags in-place (GH-23150)
  ...

encukou committed:
Remove redundant test function

encukou committed:
Merge master branch; regenerate typeslots.inc
